### PR TITLE
created a new resource to install:

### PR DESCRIPTION
**Pull request by majormoses**
called it in the recipe to avoid it breaking any existing users calling default recipe

### CODE REVIEW

**Comment by majormoses** [comment]
I verified that it does not break any of the unit tests on ubuntu (couldnt run all 71 kitchen instances on my local machine)

**Comment by bdangit** [comment]
Hi @majormoses thanks for the PR. Will review and get back.

**Comment by majormoses** [comment]
@bdangit let me know if you would like to see anything different and if you wish me to rebase the  commits together.

File: resources/install.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
+# resources/install.rb
+
+# Resource for InfluxDB install
+
+property :influxdb_key, String, default: 'https://repos.influxdata.com/influxdb.key'
+property :arch_type, String, default: 'amd64'
+default_action :install
+
```
Comment (fishnix) [inline resources/install.rb]: I would recommend allowing the addition of these repos to be optional, like is done here:

https://github.com/NorthPage/telegraf-cookbook/blob/master/resources/install.rb

File: recipes/default.rb [hunk 2]
```diff
@@ -10,26 +10,14 @@
...
-    baseurl node['influxdb']['upstream_repository']
-    gpgkey 'https://repos.influxdata.com/influxdb.key'
+influxdb_install 'influxdb' do
+  if node['kernel']['machine'].include? '64'
+    arch_type 'amd64'
+  elsif node['kernel']['machine'].include? '386'
+    arch_type 'i386'
+  else
+    Chef::Log.warning "I am unable to determine arch type for: #{node['kernel']['machine']}"
   end
```
Comment (fishnix) [inline recipes/default.rb]: It makes more sense to me if you do this in the resource, not here in the recipe.

**Comment by fishnix** [comment]
Please add the appropriate chefspec matchers as well.  I haven't had a chance to run through the test-kitchen stuff yet, but this looks good.  Thanks for the PR!

File: recipes/default.rb (hunk 2 above)
Comment (majormoses) [inline recipes/default.rb]: I thought about it while implementing and was torn as I do like the idea of having someone having a specified option in the provider in the case of improper detection. I was going for max flexibility in the provider. Let me know if you would either like this removed from the recipe, moved into the provider, or left as is.

File: resources/install.rb (hunk 1 above)
Comment (majormoses) [inline resources/install.rb]: I can certainly do that I didn't because there was no concept in the cookbook of installing not from a repo so there was no reason to add it but I can add some work around this to support that in the future.

File: recipes/default.rb (hunk 2 above)
Comment (majormoses) [inline recipes/default.rb]: Maybe we have a default value of nil and in the resource we attempt to detect if nil.

Comment (fishnix) [inline recipes/default.rb]: Maybe I'm missing something, but it seems like a problem with chef-client if the detection is wrong and I would think lots of other things would break in that case as well, just like we trust in `node.platform_family`.  My preference is to have it in the provider, but I'll let @bdangit weigh in on this one.  

I also wonder if there's a more robust match than just an include as well?  It would be particularly nice if we could just use `node['kernel']['machine']` but it looks like that won't always match in the apt repo. 🎱

**Comment by fishnix** [comment]
Thanks for making those changes and submitting this PR!  I need to get through the test-kitchen stuff, but otherwise looks great.

File: recipes/default.rb (hunk 2 above)
Comment (bdangit) [inline recipes/default.rb]: As an abstraction for installing the package influxdb, I would prefer if all the logic can be inside the resource.  the `package` and yum_repository and apt_repository or any of the repository should detect the machine right?

File: resources/install.rb [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        key influxdb_key
+        only_if { include_repository }
+      end
+    end
+
+    package 'influxdb' do
+      version node['influxdb']['version']
+    end
+  when 'tarball'
+    # TODO: Implement me
```
Comment (bdangit) [inline resources/install.rb]: If this is TODO, we should just get rid of this type of logic altogether and simplify it.

File: resources/install.rb [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    if node.platform_family? 'rhel'
+      yum_repository 'influxdb' do
+        description 'InfluxDB Repository - RHEL \$releasever'
+        baseurl node['influxdb']['upstream_repository']
+        gpgkey influxdb_key
+        only_if { include_repository }
+      end
+    else
+      package 'apt-transport-https' do
+        action :install
```
Comment (bdangit) [inline resources/install.rb]: Is this still necessary?  Shouldn't the influxdb binary already be self-contained?

Comment (bdangit) [inline resources/install.rb]: I went through a few Test Kitchen setups mostly testing 12.7.  The changes as is work and pass functional tests.  

Before we merge this, I'd like to see if we can resolve the machine logic by letting the inline chef resources take care of that.  There are others simplifications that are nice to have but can be swayed to let these be another PR.

File: resources/install.rb (hunk 3 above)
Comment (majormoses) [inline resources/install.rb]: I agree which is why I did not originally add this. I did this as per @fishnix request to allow specifying a repo or not.

File: recipes/default.rb (hunk 2 above)
Comment (majormoses) [inline recipes/default.rb]: @bdangit I think we should handle it a bit more intelligently. I was attempting to mimic current usage (which was hard coded in the recipe only for apt_repository) and assumed there was a reason for this. I can look into this further.

Maybe we could compromise and have a helper function that we call and if we have not specified we can attempt an auto discovery.

File: resources/install.rb (hunk 4 above)
Comment (majormoses) [inline resources/install.rb]: I will still check if `apt-transport-https` is still needed, just made the assumption it was.

Comment (majormoses) [inline resources/install.rb]: @bdangit it does not require it, I will remove it.

**Comment by fishnix** [comment]
I don't feel strongly about the package logic, but I will say if you don't have it, there will be some who can't use the cookbook because they don't allow access out of their networks, etc.  We've gone through those pains with telegraf. 👍

**Comment by majormoses** [comment]
@bdangit I dont see any reason to not to keep the stubbed functionality if you think this is good now I can rebase and get it ready for merging.

**Comment by bdangit** [comment]
@fishnix I will agree to adding in the logic only if the other side is fleshed out.  Stubs = deadcode.  Deadcode can lead to errors and hard dependencies with customers if not done right.

@majormoses can you squash your commit and push the PR again?  We can do a good once over again and then thumbs as needed.

**Comment by majormoses** [comment]
@bdangit let me know if there is anything else you want changed.

**Comment by bdangit** [comment]
@majormoses verifying through test-kitchen (only for 12.7 though).

File: resources/install.rb [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+      baseurl node['influxdb']['upstream_repository']
+      gpgkey influxdb_key
+      only_if { include_repository }
+    end
+  elsif node.platform_family? 'debian'
+    # see if we should auto detect
+    unless new_resource.arch_type
+      new_resource.arch_type determine_arch_type(new_resource, node)
+    end
+    apt_repository 'influxdb' do
```
Comment (bdangit) [inline resources/install.rb]: we will need add in the `package 'apt-transport-https'` otherwise the apt_repository won't be able to download from the new influxdb repo.

File: resources/install.rb [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+      uri node['influxdb']['upstream_repository']
+      distribution node['lsb']['codename']
+      components ['stable']
+      arch_type new_resource.arch_type
```
Comment (bdangit) [inline resources/install.rb]: in the releases prior to this, the arch_type was hardcoded ref: https://github.com/bdangit/chef-influxdb/blob/master/recipes/default.rb#L26

Did the current version break your builds for debian distros?

File: resources/install.rb (hunk 5 above)
Comment (majormoses) [inline resources/install.rb]: @bdangit I am confused do you want it or not? You previously asked me to remove it. I have tested and it is currently working in kitchen. I am fine with adding it back, just checking...

https://github.com/bdangit/chef-influxdb/pull/105#discussion_r61933527

File: resources/install.rb (hunk 6 above)
Comment (majormoses) [inline resources/install.rb]: No but auto discovery is almost always better than making an assumption and hard coding

File: resources/install.rb (hunk 5 above)
Comment (bdangit) [inline resources/install.rb]: Sorry for the confusion.  Upon inspection of code, it appeared that the package wasnt needed and you had confirmed it.  However, I have been checking builds based off docker images.  From clean slate on ubuntu distros, chef converges fail.  Its complaining about not having access to https.

```
  * execute[apt-get update] action run[2016-05-06T21:22:22+00:00] ERROR: execute[apt-get update] (/tmp/kitchen/cache/cookbooks/apt/providers/repository.rb line 196) had an error: Mixlib::ShellOut::ShellCommandFailed: Expected process to exit with [0], but received '100'
       ---- Begin output of apt-get update -o Dir::Etc::sourcelist='sources.list.d/influxdb.list' -o Dir::Etc::sourceparts='-' -o APT::Get::List-Cleanup='0' ----
       STDOUT:
       STDERR: E: The method driver /usr/lib/apt/methods/https could not be found.
```

If its not too much to ask, can you please add it back?

File: resources/install.rb (hunk 6 above)
Comment (bdangit) [inline resources/install.rb]: sounds good to me.

**Comment by bdangit** [comment]
once you add the package for apt-package-https, i'll thumbs up and merge.

**Comment by majormoses** [comment]
@bdangit you got it

**Comment by bdangit** [comment]
functional tests pass for chef12.7  I'll assume it will work for all other chef clients.

👍
